daemon/src/listen_protocols.rs

```diff
@@ -0,0 +1,243 @@
+use crate::collab_settlement;
+use crate::command;
+use crate::identify;
+use crate::oracle;
+use std::collections::HashSet;
+use xtra::message_channel::MessageChannel;
+use xtra::Address;
+use xtra_libp2p::NewInboundSubstream;
+use xtra_libp2p_ping::pong;
+
+pub const MAKER_LISTEN_PROTOCOLS: MakerListenProtocols = MakerListenProtocols::new(
+    xtra_libp2p_ping::PROTOCOL_NAME,
+    identify::PROTOCOL,
+    rollover::v_1_0_0::PROTOCOL,
+    rollover::v_2_0_0::PROTOCOL,
+    collab_settlement::PROTOCOL,
+);
+
+pub const TAKER_LISTEN_PROTOCOLS: TakerListenProtocols = TakerListenProtocols::new(
+    xtra_libp2p_ping::PROTOCOL_NAME,
+    identify::PROTOCOL,
+    xtra_libp2p_offer::PROTOCOL_NAME,
+);
+
+pub const REQUIRED_MAKER_LISTEN_PROTOCOLS: RequiredMakerListenProtocols =
+    RequiredMakerListenProtocols::new(
+        xtra_libp2p_ping::PROTOCOL_NAME,
+        identify::PROTOCOL,
+        rollover::v_2_0_0::PROTOCOL,
+        collab_settlement::PROTOCOL,
+    );
+
+/// Verify if the listen protocols that the `maker` supports are
+/// sufficient to fulfil the `requirements` of the taker.
+pub fn does_maker_satisfy_taker_needs(
+    maker: &HashSet<String>,
+    requirements: RequiredMakerListenProtocols,
+) -> Result<(), HashSet<String>> {
+    let missing_protocols = maker
+        .difference(&HashSet::<String>::from(requirements))
+        .cloned()
+        .collect::<HashSet<_>>();
+
+    if !missing_protocols.is_empty() {
+        return Err(missing_protocols);
+    }
+
+    Ok(())
+}
+
+/// The set of protocols that the maker's `Endpoint` is listening for.
+#[derive(Clone, Copy)]
+pub struct MakerListenProtocols {
+    ping_v1: &'static str,
+    identify_v1: &'static str,
+    rollover_v1: &'static str,
+    rollover_v2: &'static str,
+    collaborative_settlement_v1: &'static str,
+}
+
+impl MakerListenProtocols {
+    pub const fn new(
+        ping_v1: &'static str,
+        identify_v1: &'static str,
+        rollover_v1: &'static str,
+        rollover_v2: &'static str,
+        collaborative_settlement_v1: &'static str,
+    ) -> Self {
+        Self {
+            ping_v1,
+            identify_v1,
+            rollover_v1,
+            rollover_v2,
+            collaborative_settlement_v1,
+        }
+    }
+
+    /// Construct a map of protocol identifiers to actor addresses.
+    ///
+    /// This is used so that the `Endpoint` knows who to delegate to
+    /// when receiving new inbound substreams.
+    pub fn inbound_substream_handlers<R1, R2>(
+        &self,
+        ping_v1: Address<pong::Actor>,
+        identify_v1: Address<identify::listener::Actor>,
+        rollover_v1: Address<
+            rollover::v_1_0_0::maker::Actor<command::Executor, oracle::AnnouncementsChannel, R1>,
+        >,
+        rollover_v2: Address<
+            rollover::v_2_0_0::maker::Actor<command::Executor, oracle::AnnouncementsChannel, R2>,
```

Comment:
@luckysori just to make sure this is correctly rebased: The `oracle::AnnouncementsChannel` channel for rollover `v1` and `v2` is the same, correct?

Reply:
Yep. Both protocols communicate with the maker's `oracle::Actor`, and we only have one `oracle::Actor` at a time!

Reply:
That means the bound for `O` in the `rollover::...version...::maker::actor` was introduced because we don't want to have a dependency on deamon?
I felt it was needed for being generic initially, but for both the oracle actor and the executor it is purely to avoid dependencies. 
Only for the `Rates` we have two implementations that require to be generic.

Reply:
All of the trait bounds were introduced so that the protocol could be extracted into a separate crate without depending on `daemon`. Depending on `daemon` would defeat the purpose of extracting it into a separate crate and it would create a circular dependency!

But the trait bounds are also necessary if you want to test the protocol in isolation, even if you kept the protocol inside `daemon`.